include/nbl/builtin/hlsl/emulated/emulated_float64_t_utils.hlsl

```diff
@@ -0,0 +1,459 @@
+
+        return output;
+    }
+
+    type operator+(float rhs)
+    {
+        type output;
+        EmulatedType rhsAsEF64 = EmulatedType::create(rhs);
+        output.x = x + rhsAsEF64;
+        output.y = y + rhsAsEF64;
+
+        return output;
+    }
+
+    type operator+(EmulatedType rhs)
+    {
+        type output;
+        output.x = x + rhs;
+        output.y = y + rhs;
+
+        return output;
+    }
+
+    type operator+(type rhs)
+    {
+        type output;
+        output.x = x + rhs.x;
+        output.y = y + rhs.y;
+
+        return output;
+    }
+
+    type operator-(float rhs)
+    {
+        return create(x, y) + (-rhs);
+    }
+
+    type operator-(EmulatedType rhs)
+    {
+        return create(x, y) + (rhs.flipSign());
+    }
+
+    type operator-(type rhs)
+    {
+        rhs.x = rhs.x.flipSign();
+        rhs.y = rhs.y.flipSign();
+        return create(x, y) + rhs;
+    }
+
+    type operator*(float rhs)
+    {
+        type output;
+        EmulatedType rhsAsEF64 = EmulatedType::create(rhs);
+        output.x = x * rhsAsEF64;
+        output.y = y * rhsAsEF64;
+
+        return output;
+    }
+
+    type operator*(EmulatedType rhs)
+    {
+        type output;
+        output.x = x * rhs;
+        output.y = y * rhs;
+
+        return output;
+    }
+
+    type operator*(type rhs)
+    {
+        type output;
+        output.x = x * rhs.x;
+        output.y = y * rhs.y;
+
+        return output;
+    }
+};
+
+template<typename EmulatedType>
+struct emulated_vector<EmulatedType, 3>
+{
+    using type = emulated_vector<EmulatedType, 3>;
+
+    EmulatedType x;
+    EmulatedType y;
+    EmulatedType z;
+
+    EmulatedType calcComponentSum() NBL_CONST_MEMBER_FUNC
+    {
+        return x + y + z;
+    }
+
+    type operator*(NBL_CONST_REF_ARG(type) rhs) NBL_CONST_MEMBER_FUNC
+    {
+        type output;
+        output.x = x * rhs.x;
+        output.y = y * rhs.y;
+        output.z = z * rhs.z;
+
+        return output;
+    }
+};
+
+template<typename EmulatedType>
+struct emulated_vector<EmulatedType, 4>
+{
+    using type = emulated_vector<EmulatedType, 4>;
+
+    EmulatedType x;
+    EmulatedType y;
+    EmulatedType z;
+    EmulatedType w;
+};
+
+template<typename EmulatedType>
+using emulated_vector_t2 = emulated_vector<EmulatedType, 2>;
+template<typename EmulatedType>
+using emulated_vector_t3 = emulated_vector<EmulatedType, 3>;
+template<typename EmulatedType>
+using emulated_vector_t4 = emulated_vector<EmulatedType, 4>;
+
+namespace impl
+{
+
+template<typename T, typename U, uint32_t Dim>
+struct static_cast_helper<vector<T,Dim>,emulated_vector<U,Dim>,void>
+{
+    static inline vector<T,Dim> cast(emulated_vector<U,Dim> vec)
+    {
+        return vector<T,Dim>(_static_cast<T,U>(vec.x), _static_cast<T,U>(vec.y));
+    }
+};
+
+}
+
+//template<typename EmulatedType, uint32_t N, uint32_t M>
+//struct emulated_matrix_base
+//{
+//    using vec_t = emulated_vector<EmulatedType, N>;
+//    vec_t columns[M];
```

Comment:
Last I checked HLSL stores matrices using rows unlike GLSL/GLM, this is why we had to use private inheritance from `glm::mat` for our `nbl::hlsl::matrix` and override how `operator[]` works there

(@Erfan-Ahmadi correct me if I'm wrong)